### PR TITLE
Single-sided circuit breaker checks

### PR DESCRIPTION
**Pull request by EndymionJkb**
Adds single sided checks that will be needed for tokenSwap checks in ManagedPool, as well as the infrastructure (e.g., new error message).

### CODE REVIEW

File: pkg/pool-weighted/contracts/managed/CircuitBreakerStorageLib.sol [hunk 1]
```diff
@@ -218,6 +218,54 @@ library CircuitBreakerStorageLib {
...
+                _ADJUSTED_BOUND_WIDTH
+            )
+                .decompress(_ADJUSTED_BOUND_WIDTH, _MAX_BOUND_PERCENTAGE);
+        } else {
+            // Something has changed - either the weight of the token, or the composition of the pool, so we must
+            // retrieve the raw percentage bounds and do the full calculation. Decompress the bounds by shifting left.
+            boundRatio = CircuitBreakerLib.calcAdjustedBound(bound, currentWeight, isLowerBound);
+        }
+
+        // Use the ratios retrieved (or computed) above to convert raw percentage bounds to BPT price bounds.
```
Comment (nventuro) [inline pkg/pool-weighted/contracts/managed/CircuitBreakerStorageLib.sol:265]:
```suggestion
        // Use the adjusted bounds (either cached or computed) to calculate the BPT price bounds.
```

File: pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol [hunk 2]
```diff
@@ -87,4 +115,36 @@ library CircuitBreakerLib {
...
+     */
+    function calcAdjustedBound(
+        uint256 bound,
+        uint256 weight,
+        bool isLowerBound
+    ) internal pure returns (uint256 boundRatio) {
+        uint256 weightComplement = weight.complement();
+
+        // To be conservative and protect LPs, round up for the lower bound, and down for the upper bound.
+        boundRatio = (isLowerBound ? FixedPoint.powUp : FixedPoint.powDown)(bound, weightComplement);
```
Comment (TomAFrench) [inline pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol:133]:
```suggestion
        // To be conservative and protect LPs, round up for the lower bound, and down for the upper bound.
        boundRatio = (isLowerBound ? FixedPoint.powUp : FixedPoint.powDown)(bound,  weight.complement());
```

File: pkg/pool-weighted/contracts/managed/CircuitBreakerStorageLib.sol [hunk 3]
```diff
@@ -218,6 +218,54 @@ library CircuitBreakerStorageLib {
...
+        uint256 bound = circuitBreakerState.decodeUint(
+            isLowerBound ? _LOWER_BOUND_OFFSET : _UPPER_BOUND_OFFSET,
+            _BOUND_WIDTH
+        ) << _BOUND_SHIFT_BITS;
+
+        if (bound == 0) {
+            return 0;
+        }
+
+        // Retrieve the weight complement passed in and bptPrice computed when the circuit breaker was set.
```
Comment (TomAFrench) [inline pkg/pool-weighted/contracts/managed/CircuitBreakerStorageLib.sol:244]:
```suggestion
        // Retrieve the weight and bptPrice passed in computed when the circuit breaker was set.
```

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol [hunk 4]
```diff
@@ -186,6 +186,10 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, IControlled
         return _tokenState[token];
     }
 
+    function _getCircuitBreakerState(IERC20 token) internal view returns (bytes32) {
+        return _circuitBreakerState[token];
+    }
+
```
Comment (TomAFrench) [inline pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:192]: This seems to be unused.

File: pkg/pool-weighted/test/foundry/CircuitBreakerLib.t.sol [hunk 5]
```diff
@@ -118,6 +125,14 @@ contract CircuitBreakerLibTest is Test {
             uint256(bptPrice).mulUp(expectedUpperBptPrice),
             _MAX_RELATIVE_ERROR
         );
+
+        // Single-sided
+
```
Comment (TomAFrench) [inline pkg/pool-weighted/test/foundry/CircuitBreakerLib.t.sol:130]:
```suggestion
```

File: pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol [hunk 6]
```diff
@@ -23,6 +23,34 @@ import "@balancer-labs/v2-solidity-utils/contracts/math/FixedPoint.sol";
...
+     * @notice Single-sided check for whether a lower or upper circuit breaker would trip in the given pool state.
+     * @dev Compute the current BPT price from the input parameters, and compare it to the given bound to determine
+     * whether the given post-operation pool state is within the circuit breaker bounds.
+     * @param virtualSupply - the post-operation totalSupply (including protocol fees, etc.)
+     * @param weight - the normalized weight of the token we are checking.
+     * @param balance - the post-operation token balance (including swap fees, etc.). It must be an 18-decimal
+     * floating point number, adjusted by the scaling factor of the token.
+     * @param boundBptPrice - the BPT price at the limit (lower or upper) of the allowed trading range.
+     * @param isLowerBound - true if the boundBptPrice represents the lower bound.
+     * @return - boolean flag set to true if the breaker should be tripped
```
Comment (TomAFrench) [inline pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol:36]:
```suggestion
     * @return - boolean flag for whether the breaker has been tripped.
```

**Review by nventuro — APPROVED** [review]

**Review by nventuro — APPROVED** [review]

**Pull request merged**